cache/jemalloc_nodump_allocator.cc

```diff
@@ -0,0 +1,135 @@
+  assert(arena_index != 0);
+}
+
+void* JemallocNodumpAllocator::Allocate(size_t size) {
+  return mallocx(size, flags_);
+}
+
+void JemallocNodumpAllocator::Deallocate(void* p) { dallocx(p, flags_); }
+
+void* JemallocNodumpAllocator::Alloc(extent_hooks_t* extent, void* new_addr,
+                                     size_t size, size_t alignment, bool* zero,
+                                     bool* commit, unsigned arena_ind) {
+  extent_alloc_t* original_alloc =
+      original_alloc_.load(std::memory_order_relaxed);
+  assert(original_alloc != nullptr);
+  void* result = original_alloc(extent, new_addr, size, alignment, zero, commit,
+                                arena_ind);
+  if (result != nullptr) {
+    int ret = madvise(result, size, MADV_DONTDUMP);
+    if (ret != 0) {
+      fprintf(
+          stderr,
+          "JemallocNodumpAllocator failed to set MADV_DONTDUMP, error code: %d",
+          ret);
+      assert(false);
+    }
+  }
+  return result;
+}
+
+JemallocNodumpAllocator::~JemallocNodumpAllocator() {
+  assert(arena_index_ != 0);
+  std::string key = "arena." + ToString(arena_index_) + ".destroy";
+  int ret = mallctl(key.c_str(), nullptr, 0, nullptr, 0);
+  if (ret != 0) {
+    fprintf(stderr, "Failed to destroy jemalloc arena, error code: %d\n", ret);
+  }
+}
+
+size_t JemallocNodumpAllocator::UsableSize(void* p,
+                                           size_t /*allocation_size*/) const {
+  return malloc_usable_size(static_cast<void*>(p));
+}
+
+Status JemallocNodumpAllocatorFactory::NewCacheAllocator(
+    std::unique_ptr<CacheAllocator>* cache_allocator) {
+  assert(cache_allocator != nullptr);
+
+  // Create arena.
+  unsigned arena_index = 0;
+  size_t arena_index_size = sizeof(arena_index);
+  int ret =
+      mallctl("arenas.create", &arena_index, &arena_index_size, nullptr, 0);
+  if (ret != 0) {
+    return Status::Incomplete("Failed to create jemalloc arena, error code: " +
+                              ToString(ret));
+  }
+  assert(arena_index != 0);
+  int flags = MALLOCX_ARENA(arena_index) | MALLOCX_TCACHE_NONE;
+  std::string key = "arena." + ToString(arena_index) + ".extent_hooks";
+
+  // Read existing hooks.
+  extent_hooks_t* hooks;
+  size_t hooks_size = sizeof(hooks);
+  ret = mallctl(key.c_str(), &hooks, &hooks_size, nullptr, 0);
+  if (ret != 0) {
+    std::string msg =
+        "Failed to read existing hooks, error code: " + ToString(ret);
+    return Status::Incomplete("Failed to read existing hooks, error code: " +
+                              ToString(ret));
+  }
+
+  // Store existing alloc.
+  extent_alloc_t* original_alloc = hooks->alloc;
+  extent_alloc_t* expected = nullptr;
+  bool success __attribute__((__unused__)) =
+      JemallocNodumpAllocator::original_alloc_.compare_exchange_strong(
+          expected, original_alloc);
+  assert(success || original_alloc == expected);
+
+  // Set the custom hook.
+  std::unique_ptr<extent_hooks_t> new_hooks(new extent_hooks_t(*hooks));
```

Comment:
It looks we expect there to be only 1 hooks globally? If so a static one (maybe guard this part along with the compare_exchange above) might be better, since the lifetime management for hooks can get tricky.

Reply:
I would like to have separate hooks for each allocator instance, but I didn't find a way to pass non-static member method to `mallocx` as new hook. Any suggestion?

Reply:
Not sure I totally follow. Having a extent_hooks_t member in the allocator class should work? Again we will need to ensure the hooks_t is destructed after destroying the arena.

Reply:
Sorry, I didn't explain the problem well. Since `extent_hooks_t` is a typedef of function pointer,  I can only use a static method (`JemallocNodumpAllocator::Alloc`) as the hook. From the static method it is not able to access non-static members. That's why I'm forced to make `original_alloc_` static. But if there's a way to make it non-static, I'll try.

Reply:
And yes I'm holding the `extent_hooks_t` as non-static member in the class and it will get destruct after arena destruct, by destructor order. So it should work?